Functions + conjunctions + TR debug

maxaalexeeva opened the pull request:
Merged a couple of branches @alicekwak and I have been working on. 

Major changes:
- initial function rules and tests 
- can now test conjunction extractions with discontinuous text (added a method to get the text of mention arguments taking into account the discontinuous span attachment), so now there are passing conjunctions tests
- handling another type of conjunctions (no span overlap in conjoined definitions---calling these `ConjDefinitionType2` for now)
- some new filtering methods to improve quality of extractions
- major work on documenting issues in test definitions - some of the issues fixed

Minor updates:
- [update] a lot of various TR debugging
- updating some examples for rules
- adding the check from looksLikeAVariableWithGreek to the regular looksLikeAVariable
- rule and test updates
- deleting K from the list of units (we don't want to lose variables called K)

maxaalexeeva commented on automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/entities/grammar/entities.yml:
```diff
@@ -20,13 +22,13 @@ rules:
-    type: token
-    pattern: |
-      [chunk='B-VP' & word = /[A-Z]/]
+  # note: error? (repetition) 
```
@alicekwak a quick question: what was this a repetition of you thought? I can't find it. Thanks!

alicekwak replied:
Oops, I'm not sure why, but I had one more rule that is exactly the same as single_capital_VP rule. So I marked one of them off and made a note saying that the rule is a repetition. However, it seems like there's no repetition here, so I think we should enable the rule again. Sorry for the confusion!

maxaalexeeva replied:
@alicekwak, thanks! No worries!

maxaalexeeva commented:
Sorry, @pauldhein. This kind of exploded. But there are some great improvements! 

@alicekwak, I am done making changes in this branch (except for whatever feedback @pauldhein will have), so it's safe to have a look if you are interested.

pauldhein commented on automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala:
```diff
@@ -163,36 +210,104 @@ class OdinActions(val taxonomy: Taxonomy, expansionHandler: Option[ExpansionHand
+    val argsOfGivenType = mention.arguments(argType)
+    val groupedByTokenInt = groupByTokenOverlap(argsOfGivenType)
+    // if there is an overlap in args, those will be grouped => the number of groups will be lower than the number of args
+    if (groupedByTokenInt.keys.toList.length == argsOfGivenType.length) true else false
```
@maxaalexeeva the following code snippet:
```python
if <condition>:
    return True
else:
    return False
```
... can always be simplified to:
```python
return <condition>
```

I think this line follows that pattern. Can you please do the simplification?

maxaalexeeva replied:
@pauldhein, right! Thanks! Pushed! *can even not say 'return'!

pauldhein reviewed (changes requested):
This looks great @maxaalexeeva! I only have one small change request and then this is ready to merge.

maxaalexeeva commented:
@pauldhein, hi! Sorry, but I made another tiny change here https://github.com/ml4ai/automates/pull/206/commits/2781a5b61208878b2dd2b895cd32f6c8915e2bc1 (not much new for you to review). Definitely ready for final approval now (I have addressed your feedback).

pauldhein reviewed (approved):
These changes look great @maxaalexeeva! Thank you for addressing my change requests. Let's get this branch merged!

maxaalexeeva commented:
@pauldhein, yay! Thanks!
